sway-core/src/asm_generation/fuel/optimizations/symbolic_interpretation.rs

```diff
@@ -0,0 +1,232 @@
+            KnownRegValue::Eq(VirtualRegister::Constant(ConstantRegister::One)) => Some(1),
+            KnownRegValue::Eq(_) => None,
+        }
+    }
+
+    /// Check if the value depends on value of another register.
+    fn depends_on(&self, reg: &VirtualRegister) -> bool {
+        match self {
+            KnownRegValue::Const(_) => false,
+            KnownRegValue::Eq(v) => v == reg,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+struct KnownValues {
+    values: FxHashMap<VirtualRegister, KnownRegValue>,
+}
+
+impl KnownValues {
+    /// Resolve a register to a known value.
+    fn resolve(&self, v: &VirtualRegister) -> Option<KnownRegValue> {
+        match v {
+            VirtualRegister::Constant(ConstantRegister::Zero) => Some(KnownRegValue::Const(0)),
+            VirtualRegister::Constant(ConstantRegister::One) => Some(KnownRegValue::Const(1)),
+            other => self.values.get(other).cloned(),
+        }
+    }
+
+    /// Clear values that depend on a register having a specific value.
+    fn clear_dependent_on(&mut self, reg: &VirtualRegister) {
+        self.values.retain(|_, v| !v.depends_on(reg));
+    }
+
+    /// Insert a known value for a register.
+    fn assign(&mut self, dst: VirtualRegister, value: KnownRegValue) {
+        self.clear_dependent_on(&dst);
+        self.values.insert(dst, value);
+    }
+}
+
+/// What knowledge is lost after an op we don't know how to interpret?
+#[derive(Clone, Debug)]
+enum ResetKnown {
+    /// Reset all known values
+    Full,
+    /// Reset non-virtual registers in addition to defs
+    NonVirtual,
+    /// Only the `def_registers` and `def_const_registers` are reset
+    Defs,
+}
+impl ResetKnown {
+    fn apply(&self, op: &Op, known_values: &mut KnownValues) {
+        match self {
+            ResetKnown::Full => {
+                known_values.values.clear();
+            }
+            ResetKnown::NonVirtual => {
+                Self::Defs.apply(op, known_values);
+                known_values
+                    .values
+                    .retain(|k, _| matches!(k, VirtualRegister::Virtual(_)));
+            }
+            ResetKnown::Defs => {
+                for d in op.def_registers() {
+                    known_values.clear_dependent_on(d);
+                    known_values.values.remove(d);
+                }
+                for d in op.def_const_registers() {
+                    known_values.clear_dependent_on(d);
+                    known_values.values.remove(d);
+                }
+            }
+        }
+    }
+}
+
+impl AbstractInstructionSet {
+    /// Symbolically interpret code and propagate known register values.
+    pub(crate) fn interpret_propagate(mut self) -> AbstractInstructionSet {
+        if self.ops.is_empty() {
+            return self;
+        }
+
+        // The set of labels that are jump targets
+        // todo: build proper control flow graph instead
+        let jump_target_labels: FxHashSet<Label> = self
```

Comment:
We could track `FxHashSet<Label, u32>`, i.e., how many times `label` is a target (or in other words, how many predecessors a label has).

With that ... [see next]